File: Data Structures & Algorithms/sort-an-array/submission-1.py
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        def mergeSortHelper(l, r):
            if l >= r:
                return

            m = l + (r - l) // 2
            mergeSortHelper(l, m)
            mergeSortHelper(m + 1, r)

            mergeArrs(l, m, r)

        def mergeArrs(l, m, r):
            left = [nums[i] for i in range(l, m + 1)]
            print(left)
            right = [nums[i] for i in range(m + 1, r + 1)]
            print(right)

            leftPtr = 0
            rightPtr = 0
            currPtr = l

            while leftPtr < len(left) and rightPtr < len(right):
                if left[leftPtr] < right[rightPtr]:
                    nums[currPtr] = left[leftPtr]
                    leftPtr += 1
                else:
                    nums[currPtr] = right[rightPtr]
                    rightPtr += 1
                
                currPtr += 1
            
            while leftPtr < len(left):
                nums[currPtr] = left[leftPtr]
                leftPtr += 1
                currPtr += 1

            while rightPtr < len(right):
                nums[currPtr] = right[rightPtr]
                rightPtr += 1
                currPtr += 1

        mergeSortHelper(0, len(nums) - 1)
        return nums

File: Data Structures & Algorithms/mergeSort/submission-1.cpp
// Definition for a Pair
// class Pair {
// public:
//     int key;
//     string value;
//
//     Pair(int key, string value) : key(key), value(value) {}
// };
class Solution {
public:
    vector<Pair> mergeSort(vector<Pair>& pairs) {
        mergeSortHelper(pairs, 0, pairs.size() - 1);
        return pairs;
    }

    void mergeSortHelper(vector<Pair>& pairs, int s, int e) {
        if (s >= e) return;

        int m = s + (e - s) / 2;
        mergeSortHelper(pairs, s, m);
        mergeSortHelper(pairs, m + 1, e);

        merge(pairs, s, m, e);
    }

    void merge(vector<Pair>& pairs, int s, int m, int e) {
        vector<Pair> left(pairs.begin() + s, pairs.begin() + m + 1);
        vector<Pair> right(pairs.begin() + m + 1, pairs.begin() + e + 1);

        int l = 0;
        int r = 0;
        int i = s;

        while (l < left.size() && r < right.size()) {
            if (right[r].key < left[l].key) {
                pairs[i] = right[r];
                r++;
            } else {
                pairs[i] = left[l];
                l++;
            }

            i++;
        }

        while (l < left.size()) {
            pairs[i] = left[l];
            l++;
            i++;
        }

        while (r < right.size()) {
            pairs[i] = right[r];
            r++;
            i++;
        }
    }
};
